### PR TITLE
[release-4.16] OCPBUGS-34979: Updates default security context behavior for catalog source pods

### PR DESCRIPTION
**Pull request by perdasilva**
Cherry pick of https://github.com/openshift/operator-framework-olm/commit/47978e1def3daa76b07afa3f16ae9225ce70346f

    Refactor security context configuration in pod reconciler

This change updates the logic for setting security contexts within the OLM pod reconciler. Now, it differentiates between 'Restricted' and 'Legacy' security contexts more explicitly. The 'Restricted' security context applies default security settings unless overridden, while the 'Legacy' context clears all security settings. When no security context is configured, it defaults to restricted. Additionally, the related tests have been updated to reflect these changes and ensure correct behavior.

    Add checking of the namespace PSA restrictions

    Fix linter issues

    fixes

Upstream-repository: operator-lifecycle-manager
Upstream-commit: 9b2802151528e556d78b5fe55bcb8c89070efe58

### CODE REVIEW

**Review by tmshort — APPROVED** [review]
/lgtm

**Comment by jianzhangbjz** [comment]
Pre-merge test pass, details: https://issues.redhat.com/browse/OCPBUGS-34979
/lgtm
/label qe-approved
/label cherry-pick-approved

**Comment by joelanford** [comment]
This isn't a critical bug fix, so I'm holding off until we have a 4.16.0. We can get this into a 4.16.z.

File: staging/api/crds/operators.coreos.com_catalogsources.yaml [hunk 1]
```diff
@@ -989,19 +989,15 @@ spec:
...
+                        run in PSA `baseline` or `privileged` namespaces. If the SecurityContextConfig is unspecified, the mode will be
+                        determined by the namespace's PSA configuration. If the namespace is enforcing `restricted` mode, then the pod
+                        will be configured as if `restricted` was specified. Otherwise, it will be configured as if `legacy` was
+                        specified. Specifying a value other than `legacy` or `restricted` result in a validation error. When using older
+                        catalog images, which can not run in `restricted` mode, the SecurityContextConfig should be set to `legacy`.
 
 
                         More information about PSA can be found here: https://kubernetes.io/docs/concepts/security/pod-security-admission/'
                       type: string
-                      default: legacy
```
Comment (everettraven) [inline staging/api/crds/operators.coreos.com_catalogsources.yaml:1004]: Isn't this technically a breaking change? Is this something we should be concerned about?

Comment (perdasilva) [reply staging/api/crds/operators.coreos.com_catalogsources.yaml:1004]: Great catch. Kinda, we subbed the default for a probe. We look at the namespace for PSA annotations and make a decision.

Comment (perdasilva) [reply staging/api/crds/operators.coreos.com_catalogsources.yaml:1004]: That PR went ahead. It didn't include the API fixes T_T

**Pull request merged**